File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/wallet/WalletType.java
package org.hyperledger.indy.sdk.wallet;

import org.hyperledger.indy.sdk.ErrorCode;

import com.sun.jna.Callback;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

/**
 * Base type for implementing custom wallet types.
 */
public abstract class WalletType {

	/**
	 * Callback called when a wallet is being created.
	 */
	private Callback createCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(String name, String config, String credentials) {

			try
			{
				return create(name, config, credentials).value();
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called when a wallet is being opened.
	 */
	private Callback openCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(String name, String config, String runtime_config, String credentials, Pointer handle) {

			try
			{
				HandleByReference walletHandle = new HandleByReference(); 
				ErrorCode result = open(name, config, runtime_config, credentials, walletHandle);
				handle.setInt(0, walletHandle.getValue());
				
				return result.value();
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called when a value is being set on a wallet.
	 */
	private Callback setCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int handle, String key, String value) {
			
            try
			{
            	CustomWallet wallet = getWalletByHandle(handle);
                return wallet.set(key, value).value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called value is requested from a wallet.
	 */
	private Callback getCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int handle, String key, PointerByReference value_ptr) {

			try
			{
            	CustomWallet wallet = getWalletByHandle(handle);
            	            	
            	StringByReference resultString = new StringByReference();
                ErrorCode result = wallet.get(key, resultString);
                
                if(result != ErrorCode.Success)
                	return result.value();
                
                Pointer marshalledValue = marshalToNative(resultString.getValue());
                value_ptr.setValue(marshalledValue);
                		
                return result.value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called when an unexpired value is being requested from a wallet.
	 */
	private Callback getNotExpiredCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int handle, String key, PointerByReference value_ptr) {
			
			try
			{
            	CustomWallet wallet = getWalletByHandle(handle);
            	
            	StringByReference resultString = new StringByReference();
                ErrorCode result = wallet.getNotExpired(key, resultString);
                
                if(result != ErrorCode.Success)
                	return result.value();
                
                Pointer marshalledValue = marshalToNative(resultString.getValue());
                value_ptr.setValue(marshalledValue);
                		
                return result.value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called when a list of values is being requested from a wallet.
	 */
	private Callback listCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int handle, String key_prefix, PointerByReference values_json_ptr) {

			try
			{
            	CustomWallet wallet = getWalletByHandle(handle);
            	
            	StringByReference resultString = new StringByReference();
                ErrorCode result = wallet.list(key_prefix, resultString);
                
                if(result != ErrorCode.Success)
                	return result.value();
                
                Pointer marshalledValue = marshalToNative(resultString.getValue());
                values_json_ptr.setValue(marshalledValue);
                		
                return result.value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}
		}
	};

	/**
	 * Callback called when a wallet is being closed.
	 */
	private Callback closeCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int handle) {

			try
			{            	
            	return close(handle).value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}			
		}
	};

	/**
	 * Callback called when a wallet is being deleted.
	 */
	private Callback deleteCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(String name, String config, String credentials) {

			try
			{            	
            	return delete(name, config, credentials).value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}			
		}
	};

	/**
	 * Callback called when value requested from a wallet is no longer being used and should be freed.
	 */
	private Callback freeCb = new Callback() {

		@SuppressWarnings("unused")
		public int callback(int wallet_handle, Pointer value) {

			try
			{       			
				Native.free(Pointer.nativeValue(value));
            	return ErrorCode.Success.value();  
			}
			catch(Exception e)
			{
				return ErrorCode.CommonInvalidState.value();
			}			
		}
	};
	
	/**
	 * Gets the create callback.
	 * 
	 * @return The create callback.
	 */
	Callback getCreateCb() {
		return createCb;
	}

	/**
	 * Gets the open callback.
	 * 
	 * @return The open callback.
	 */
	Callback getOpenCb() {
		return openCb;
	}

	/**
	 * Gets the set callback.
	 * 
	 * @return The set callback.
	 */
	Callback getSetCb() {
		return setCb;
	}

	/**
	 * Gets the get callback.
	 * 
	 * @return The get callback.
	 */
	Callback getGetCb() {
		return getCb;
	}

	/**
	 * Gets the getNotExpired callback.
	 * 
	 * @return The getNotExpired callback.
	 */
	Callback getGetNotExpiredCb() {
		return getNotExpiredCb;
	}

	/**
	 * Gets the list callback.
	 * 
	 * @return The list callback.
	 */
	Callback getListCb() {
		return listCb;
	}

	/**
	 * Gets the close callback.
	 * 
	 * @return The close callback.
	 */
	Callback getCloseCb() {
		return closeCb;
	}

	/**
	 * Gets the delete callback.
	 * 
	 * @return The delete callback.
	 */
	Callback getDeleteCb() {
		return deleteCb;
	}

	/**
	 * Gets the free callback.
	 * 
	 * @return The free callback.
	 */
	Callback getFreeCb() {
		return freeCb;
	}
	
	/**
	 * Creates a new wallet.
	 * 
	 * @param name The name of the wallet to create.
	 * @param config The configuration of the wallet.
	 * @param credentials The credentials of the wallet.
	 * @return An Errorcode indicating the outcome.
	 */
	public abstract ErrorCode create(String name, String config, String credentials);
	
	/**
	 * Opens a wallet.
	 * 
	 * @param name The name of the wallet to open.
	 * @param config The configuration for the wallet specified on creation.
	 * @param runtimeConfig The runtime configuration for the wallet.
	 * @param credentials The credentials for the wallet.
	 * @param walletHandle A handle for the opened wallet instance to be set by the implementer.
	 * @return An Errorcode indicating the outcome.
	 */
	public abstract ErrorCode open(String name, String config, String runtimeConfig, String credentials, HandleByReference walletHandle);
		
	/**
	 * Closes a wallet.
	 * 
	 * @param handle The handle of the wallet to close.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode close(int handle);
	
	/**
	 * Deletes a wallet.
	 * 
	 * @param name The name of the wallet.
	 * @param config The configuration of the wallet.
	 * @param credentials The credentials of the wallet.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode delete(String name, String config, String credentials);
		
	/**
	 * Gets an open wallet by its handle.
	 * 
	 * @param handle The handle of the open wallet.
	 * @return The wallet instance associated with the handle.
	 */
	protected abstract CustomWallet getWalletByHandle(int handle);
	
	/**
	 * Marshals a string value to unmanaged memory and returns a pointer to the native memory.
	 * 
	 * @param value The value to marshal.
	 * @return A pointer to the native memory containing the marshalled value.
	 */
	private Pointer marshalToNative(String value) {
		byte[] bytes = Native.toByteArray(value, "UTF-8");
		Pointer pointer = new Pointer(Native.malloc(bytes.length));
		pointer.write(0, bytes, 0, bytes.length);
		return pointer;
	}
	
	/**
	 * A result returned from a wallet method that gets a value.
	 */
	public class StringByReference {
		
		private String value;
		
		/**
		 * Sets the value for the result.
		 * 
		 * @param value The value.
		 */
		public void setValue(String value) {
			this.value = value;
		}
		
		/**
		 * Gets the value from the result.
		 * 
		 * @return The value.
		 */
		public String getValue() {
			return this.value;
		}		
	}
	
	/**
	 * A handle returned from a wallet method.
	 */
	public class HandleByReference {
		private int handle;
		
		/**
		 * Sets the handle value.
		 * 
		 * @param value The handle.
		 */
		public void setValue(int value) {
			handle = value;
		}
		
		/**
		 * Gets the handle value.
		 * 
		 * @return The handle.
		 */
		public int getValue() {
			return handle;
		}
	}
	
}

File: wrappers/dotnet/indy-sdk-dotnet/WalletApi/WalletType.cs
using Hyperledger.Indy.Utils;
using System;
using System.Runtime.InteropServices;
using System.Text;
using static Hyperledger.Indy.WalletApi.NativeMethods;

namespace Hyperledger.Indy.WalletApi
{
    /// <summary>
    /// Base type for implementing custom wallet types.
    /// </summary>
    public abstract class WalletType
    {
        /// <summary>
        /// Initializes a new WalletType.
        /// </summary>
        protected WalletType()
        {
            CreateCallback = CreateHandler;
            OpenCallback = OpenHandler;
            SetCallback = SetHandler;
            GetCallback = GetHandler;
            GetNotExpiredCallback = GetNotExpiredHandler;
            ListCallback = ListHandler;
            CloseCallback = CloseHandler;
            DeleteCallback = DeleteHandler;
            FreeCallback = FreeHandler;
        }

        /// <summary>
        /// The delegate to call when a wallet is being created.
        /// </summary>
        internal WalletTypeCreateDelegate CreateCallback { get; }

        /// <summary>
        /// The delegate to call when a wallet is being opened.
        /// </summary>
        internal WalletTypeOpenDelegate OpenCallback { get; }

        /// <summary>
        /// The delegate to call when a value is set  on a wallet.
        /// </summary>
        internal WalletTypeSetDelegate SetCallback { get; }

        /// <summary>
        /// The delegate to call when a value is requested from a wallet.
        /// </summary>
        internal WalletTypeGetDelegate GetCallback { get; }

        /// <summary>
        /// The delegate to call when an unexpired value is requested from a wallet.
        /// </summary>
        internal WalletTypeGetNotExpiredDelegate GetNotExpiredCallback { get; }

        /// <summary>
        /// The delegate to call when a list of values is requested from a wallet.
        /// </summary>
        internal WalletTypeListDelegate ListCallback { get; }

        /// <summary>
        /// The delegate to call when a wallet is being closed.
        /// </summary>
        internal WalletTypeCloseDelegate CloseCallback { get; }

        /// <summary>
        /// The delegate to call when a wallet is being deleted.
        /// </summary>
        internal WalletTypeDeleteDelegate DeleteCallback { get; }

        /// <summary>
        /// The delegate to call when a value returned by a wallet is to be freed.
        /// </summary>
        internal WalletTypeFreeDelegate FreeCallback { get; }

        /// <summary>
        /// Gets a wallet by its handle.
        /// </summary>
        /// <param name="walletHandle">The handle of the wallet.</param>
        /// <returns>The wallet instance associated with the handle.</returns>
        protected abstract ICustomWallet GetWalletByHandle(int walletHandle);

        /// <summary>
        /// Handler for wallet creation.
        /// </summary>
        /// <param name="name">The name of the wallet to create.</param>
        /// <param name="config">The configuration for the wallet.</param>
        /// <param name="credentials">The credentials of the wallet.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode CreateHandler(string name, string config, string credentials)
        {
            ParamGuard.NotNullOrWhiteSpace(name, "name");

            try
            {
                return Create(name, config, credentials);

            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }            
        }

        /// <summary>
        /// Handler for opening a wallet.
        /// </summary>
        /// <param name="name">The name of the wallet to open.</param>
        /// <param name="config">The wallet configuration that was registered on creation.</param>
        /// <param name="runtimeConfig">The runtime configuration to use for the wallet.</param>
        /// <param name="credentials">The wallet credentials.</param>
        /// <param name="handle">A handle returned for the opened wallet instance.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode OpenHandler(string name, string config, string runtimeConfig, string credentials, ref int handle)
        {
            ParamGuard.NotNullOrWhiteSpace(name, "name");

            try
            {
                return Open(name, config, runtimeConfig, credentials, out handle);
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for setting a value on an opened wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <param name="key">The key of the value to set.</param>
        /// <param name="value">The value to set.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode SetHandler(int handle, string key, string value)
        {
            ParamGuard.NotNullOrWhiteSpace(key, "key");

            try
            {
                var wallet = GetWalletByHandle(handle);
                return wallet.Set(key, value);                
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for getting a value from an opened wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <param name="key">The key of the value to get.</param>
        /// <param name="value_ptr">The returned pointer to the value.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode GetHandler(int handle, string key, ref IntPtr value_ptr)
        {
            ParamGuard.NotNullOrWhiteSpace(key, "key");

            try
            {
                var wallet = GetWalletByHandle(handle);

                string value;
                var result = wallet.Get(key, out value);

                if (result != ErrorCode.Success)
                    return result;
                
                value_ptr = MarshalToUnmanaged(value);

                return ErrorCode.Success;
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for getting an unexpired value from an open wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <param name="key">The key of the value to get.</param>
        /// <param name="value_ptr">The returned pointer to the value.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode GetNotExpiredHandler(int handle, string key, ref IntPtr value_ptr)
        {
            ParamGuard.NotNullOrWhiteSpace(key, "key");

            try
            {
                var wallet = GetWalletByHandle(handle);

                string value;
                var result = wallet.GetNotExpired(key, out value);

                if (result != ErrorCode.Success)
                    return result;

                value_ptr = MarshalToUnmanaged(value);

                return ErrorCode.Success;
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for getting a list of values from an open wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <param name="keyPrefix">The prefix to filter keys by.</param>
        /// <param name="values_json_ptr">The returned pointer to the value.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode ListHandler(int handle, string keyPrefix, ref IntPtr values_json_ptr)
        {
            try
            {
                var wallet = GetWalletByHandle(handle);

                string value;
                var result = wallet.List(keyPrefix, out value);

                if (result != ErrorCode.Success)
                    return result;

                values_json_ptr = MarshalToUnmanaged(value);

                return ErrorCode.Success;
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for closing an open wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode CloseHandler(int handle)
        {
            try
            {
                return Close(handle);                
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for deleting a wallet.
        /// </summary>
        /// <param name="name">The name of the wallet to delete.</param>
        /// <param name="config">The configuration of the wallet.</param>
        /// <param name="credentials">The credentials of the wallet.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode DeleteHandler(string name, string config, string credentials)
        {
            ParamGuard.NotNullOrWhiteSpace(name, "name");

            try
            {
                return Delete(name, config, credentials);
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }
        }

        /// <summary>
        /// Handler for freeing a value returned by an open wallet instance.
        /// </summary>
        /// <param name="handle">The handle of the wallet instance.</param>
        /// <param name="value">The pointer to the value to free.</param>
        /// <returns>An ErrorCode indicating the outcome of the operation.</returns>
        private ErrorCode FreeHandler(int handle, IntPtr value)
        {
            try
            {
                var wallet = GetWalletByHandle(handle);

                Marshal.FreeHGlobal(value);                
                return ErrorCode.Success;
            }
            catch (Exception)
            {
                return ErrorCode.CommonInvalidState;
            }

        }

        /// <summary>
        /// Marshals a string to unmanaged memory.
        /// </summary>
        /// <param name="value">The string value to marshal.</param>
        /// <returns>A pointer to the unmanaged memory.</returns>
        private IntPtr MarshalToUnmanaged(string value)
        {
            byte[] buffer = Encoding.UTF8.GetBytes(value); // not null terminated
            Array.Resize(ref buffer, buffer.Length + 1);
            buffer[buffer.Length - 1] = 0; // terminating 0
            IntPtr unmanagedMemoryPtr = Marshal.AllocHGlobal(buffer.Length);
            Marshal.Copy(buffer, 0, unmanagedMemoryPtr, buffer.Length);
            return unmanagedMemoryPtr;
        }

        /// <summary>
        /// Allows an implementer to create a new wallet.
        /// </summary>
        /// <remarks>
        /// <para>When implementing a custom wallet this method is responsible for creating the new wallet
        /// and storing its configuration and credentials.
        /// </para>
        /// </remarks>
        /// <param name="name">The name of the wallet to create.</param>
        /// <param name="config">The configuration for the wallet.</param>
        /// <param name="credentials">The credentials for the wallet.</param>
        /// <returns>An <see cref="ErrorCode"/> value indicating the outcome of the operation.</returns>
        public abstract ErrorCode Create(string name, string config, string credentials);

        /// <summary>
        ///  Allows an implementer to open a wallet.
        /// </summary>
        /// <remarks>
        /// When implementing a custom wallet this method is responsible for opening the wallet and returning
        /// a handle for the opened wallet.  The value of the <paramref name="runtimeConfig"/> parameter
        /// should override any corresponding values provided in the <paramref name="config"/> parameter
        /// and value of the <paramref name="credentials"/> parameter should be used to control access 
        /// to the wallet.
        /// </remarks>
        /// <param name="name">The name of the wallet to open.</param>
        /// <param name="config">The configuration for the wallet.</param>
        /// <param name="runtimeConfig">The runtime configuration for the wallet.</param>
        /// <param name="credentials">The credentials of the wallet.</param>
        /// <param name="walletHandle">The handle assigned to the wallet.</param> 
        /// <returns>An <see cref="ErrorCode"/> value indicating the outcome of the operation.</returns>
        public abstract ErrorCode Open(string name, string config, string runtimeConfig, string credentials, out int walletHandle);

        /// <summary>
        ///  Allows an implementer to close a wallet.
        /// </summary>
        /// <remarks>
        /// When implementing a custom wallet this method is responsible for closing the wallet with 
        /// the handle allocated earlier in the <see cref="Open(string, string, string, string, out int)"/>
        /// method.
        /// </remarks>
        /// <param name="walletHandle">The handle of the wallet to close.</param>
        /// <returns>An <see cref="ErrorCode"/> value indicating the outcome of the operation.</returns>
        public abstract ErrorCode Close(int walletHandle);

        /// <summary>
        ///  Allows an implementer to delete a wallet.
        /// </summary>
        /// <remarks>
        /// When implementing a custom wallet this method is responsible for deleting a wallet created
        /// earlier via the <see cref="Create(string, string, string)"/> method.  The value of the 
        /// <paramref name="credentials"/> parameter should be used to control access whether or not
        /// the wallet can be deleted.</remarks>
        /// <param name="name">The name of the wallet being deleted</param>
        /// <param name="config">The configuration of the wallet.</param>
        /// <param name="credentials">The credentials of the wallet.</param>
        /// <returns>An <see cref="ErrorCode"/> value indicating the outcome of the operation.</returns>
        public abstract ErrorCode Delete(string name, string config, string credentials);
    }
}